std/parallelism.d

```diff
@@ -99,87 +99,47 @@ else version(NetBSD)
     version = useSysctlbyname;
 }
 
-
-version(Windows)
-{
-    // BUGS:  Only works on Windows 2000 and above.
-    shared static this()
-    {
-        import core.sys.windows.windows : SYSTEM_INFO, GetSystemInfo;
-        import std.algorithm.comparison : max;
-
-        SYSTEM_INFO si;
-        GetSystemInfo(&si);
-        totalCPUs = max(1, cast(uint) si.dwNumberOfProcessors);
-    }
-
-}
-else version(linux)
-{
-    shared static this()
-    {
-        import core.sys.posix.unistd : _SC_NPROCESSORS_ONLN, sysconf;
-        totalCPUs = cast(uint) sysconf(_SC_NPROCESSORS_ONLN);
-    }
-}
-else version(Solaris)
+/*
+A lazily initialized global constant. The underlying value is a shared global
+statically initialized to `outOfBandValue` which must not be a legit value of
+the constant. Upon the first call the situation is detected and the global is
+initialized by calling `initializer`. The initializer is assumed to be pure
+(even if not marked as such), i.e. return the same value upon repeated calls.
+For that reason, no special precautions are taken so `initializer` may be called
+more than one time leading to benign races on the cached value.
+
+In the quiescent state the cost of the function is an atomic load from a global.
+*/
+package @property @nogc nothrow pure @trusted
+T lazilyInitializedConstant(T, T outOfBandValue, alias initializer)()
 {
-    shared static this()
-    {
-        import core.sys.posix.unistd : _SC_NPROCESSORS_ONLN, sysconf;
-        totalCPUs = cast(uint) sysconf(_SC_NPROCESSORS_ONLN);
-    }
+    shared T result = outOfBandValue;
+    // Short path
+    auto local = atomicLoad(result);
+    if (local != outOfBandValue) return local;
+    // Long path
+    local = (cast(size_t function() @nogc nothrow pure @trusted) &initializer)();
```

Comment:
You should check the type of `initializer`, because it may be `delegate`, a literal value, or any function pointer type pointing to a function taking one more arguments (which should be statically disallowed).